### ISSUE TITLE
[ROADMAP] Full Migration to TypeScript

### ISSUE THREAD

**Issue by nathanlesage** [Zettlr/Zettlr#2661]
Hey everyone,

Zettlr 2.0 is out, so it's already time to work on a roadmap for the next steps. Zettlr 2.0 is a quantum leap, but there are still a few things in the codebase that annoy me, so here's what I would like to do!

First and foremost, I personally won't do anything this year anymore, because I have an important examination on Nov 30 and that one has absolute priority. And, afterwards, I'll be on deserved vacation. But after Christmas I will be getting back to those things.

That being said, there are many smaller issues still open here, and if you don't want to wait for me, please go forward: all of those issues are free for all!

**I would like to emphasize that I reserve implementation of the roadmap items I list in this specific issue for myself**. This is not because I'm a greedy despot, but rather because these require a lot of evaluation on the larger scale (where will Zettlr go in, say, five years?), plus as a main-maintainer I should know what the code is doing for the most part. That being said, I will implement these things via Pull Requests instead of directly committing to the main repository, so that all of you will have a chance to review the changes and utter your opinions!

## 1. Migrate remaining vanilla-JS files to TypeScript

First, I will migrate the remaining JavaScript files to TypeScript. I have refrained from doing so because I wanted 2.0 to first and foremost work for users, and not necessarily produce beautiful code. Now that 2.0 is out I'll be able to do the TypeScript migration fully. This requires a lot of thinking about the internal API and typings, and I will notice many, many things where we can actually combine types, etc. This is why I'd like to do this myself. Afterwards, the idea is, that it's much easier to implement everything because you will precisely know what you get when calling a function, and what the function requires.

This roadmap item will also fix every linter warning we still have (since all of them are solely due to missing typings). Except the Vue-file-warnings, because they are still in JavaScript. Which leads me to the next issue:

## ~~2. Migration to Vue.js 3 (w/ TypeScript)~~ **Done!**

The next step is to also move the Vue files to Vue 3, which coincidentally is also TypeScript compatible. So this will make sure the codebase is 100% TypeScript and additionally the Vue files will finally have typings, making accidents much less likely and the linter much more friendly.

## 3. Migrate to CodeMirror 6

After these two things have been done, we are ready for the final piece of the TypeScript migration: Going CodeMirror 6. I do that one last, first because CM6 is still not stable, and second because it's the most complex component of Zettlr. This one will likely be a behemoth to handle, so I'd like to get everything else working.

## 4. Full Backlink Implementation

This one has been demanded fiercly for the past years, so I'll finally go about this. Since this will require hefty changes to the internal API, I want to do this only after the codebase is cleaned up. There is, however, a real chance that I will pull this through *before* the CodeMirror migration.

## 5. Support for Arbitrary File Locations

The FSAL I implemented a year back was also my first attempt to prepare everything to make Zettlr completely independent of the file system. What I will do in this milestone is that I will completely outsource the calls to the local file system (using the `fs` module) into an adapter and then add more adapters, the first being a WebDav adapter. Further adapters will then cover Dropbox and OneDrive, `git`, etc. This will make working with Zettlr and your own files easier especially if you have a company computer and no, e.g., Nextcloud client installed. *(Also, you know what this enables? A potential PWA drafted directly from the source code of Zettlr so that we can perspectively have the app work not just on Desktop, but also on mobile, I heard there are demands for that. But that's explicitly NOT part of the roadmap.)*
